From 2a72901d0ee848a9f9432b4016a1c4c8136a57d4 Mon Sep 17 00:00:00 2001
From: Sven Neumann <neumann@teufel.de>
Date: Wed, 30 Sep 2015 19:57:11 +0200
Subject: [PATCH] connman: fix a potential crash

cm_service_add() creates a CMService struct with a NULL
service_proxy field and then uses the asynchronous
g_dbus_proxy_new_for_bus() to create the proxy. So the
service_proxy may be NULL when cm_service_free() is called
before the proxy is actually created.
---
 libgupnp/gupnp-connman-manager.c | 18 ++++++++++--------
 1 file changed, 10 insertions(+), 8 deletions(-)

diff --git a/libgupnp/gupnp-connman-manager.c b/libgupnp/gupnp-connman-manager.c
index afce14c..a764c80 100644
--- a/libgupnp/gupnp-connman-manager.c
+++ b/libgupnp/gupnp-connman-manager.c
@@ -235,17 +235,19 @@ cm_service_new (GUPnPConnmanManager *manager,
 static void
 cm_service_free (CMService *cm_service)
 {
-        GDBusConnection *cnx;
+        if (cm_service->proxy != NULL) {
+                GDBusConnection *cnx;
 
-        cnx = g_dbus_proxy_get_connection (cm_service->proxy);
+                cnx = g_dbus_proxy_get_connection (cm_service->proxy);
 
-        if (cm_service->sig_prop_id) {
-                g_dbus_connection_signal_unsubscribe (cnx,
-                                                      cm_service->sig_prop_id);
-                cm_service->sig_prop_id = 0;
-        }
+                if (cm_service->sig_prop_id) {
+                        g_dbus_connection_signal_unsubscribe (cnx,
+                                                              cm_service->sig_prop_id);
+                        cm_service->sig_prop_id = 0;
+                }
 
-        g_object_unref (cm_service->proxy);
+                g_object_unref (cm_service->proxy);
+        }
 
         if (cm_service->context != NULL) {
                 g_signal_emit_by_name (cm_service->manager,
-- 
2.4.3

